src/DataStructures/Python/Numpy.hpp

```diff
@@ -0,0 +1,14 @@
+// Distributed under the MIT License.
```

Comment:
What’s the purpose of this separate file? If it is to bundle the `define`s, then also use it in `Bindings.cpp`

Reply:
I believe that the `Bindings.cpp` file is unique because it initializes the Python Module,  so it cannot have the `#define NO_IMPORT_ARRAY`, as opposed to all of the other files which use the Array API, which need to have `NO_IMPORT_ARRAY` defined, see https://docs.scipy.org/doc/numpy-1.15.1/reference/c-api.array.html, although @nilsdeppe would be better equipped to answer this than me.

Reply:
The reason for originally preferring what is done now is exact what @isaaclegred said. I didn't really want:

```cpp
#define NO_IMPORT_ARRAY
#include "Numpy.hpp"
```

since this does the weird thing of controlling the code generated in `Numpy.hpp` with a local macro. This is an antipattern that will blow up when C++ modules lands and so I think should be avoided.

Reply:
makes sense 👍

Reply:
@isaaclegred could you please add a comment like:
```cpp
// We do not use the antipattern:
// #define NO_IMPORT_ARRY
// #include "DataStructures/Python/Numpy.hpp"
// because
// 1. This means we are controlling code in the header file with a
//    local macro, which can be very confusing and should generally be avoided.
// 2. When C++ modules land this type of pattern will no longer really be possible
//    since it goes against exactly what modules is trying to do: make things
//    modular.
```
to where the Bindings.cpp file?

We should probably also move `numpy.hpp` into `src/PythonBindings` so other people will find it :)

I'm fine with you squashing all this.

Reply:
Sounds good, will do.